File: go/text_normalizer.go
package textnormalizer

import (
	"regexp"
	"strings"
	"unicode"
)

// TextNormalizer is a struct for normalizing text containing unusual writing styles,
// such as repeated characters, text in asterisks, or unusual capitalization.
type TextNormalizer struct{}

// NewTextNormalizer creates a new instance of TextNormalizer.
func NewTextNormalizer() *TextNormalizer {
	return &TextNormalizer{}
}

// NormalizeText normalizes text by applying multiple normalization rules.
func (tn *TextNormalizer) NormalizeText(input string) string {
	if input == "" {
		return input
	}

	result := input

	// Normalize stretched words (e.g., "hellooo" -> "hello")
	result = tn.NormalizeRepeatedCharacters(result)

	// Remove text in asterisks (e.g., "*cough*")
	result = tn.RemoveAsterisks(result)

	// Normalize capitalization (e.g., "HeLLo" -> "Hello")
	result = tn.NormalizeCasing(result)

	return result
}

// NormalizeRepeatedCharacters normalizes repeated characters in text.
func (tn *TextNormalizer) NormalizeRepeatedCharacters(input string) string {
	if input == "" {
		return input
	}

	// Define word boundaries to identify words
	re := regexp.MustCompile(`\b\w+\b`)

	return re.ReplaceAllStringFunc(input, func(word string) string {
		var normalized strings.Builder
		var lastChar rune
		var isFirst = true

		for _, char := range word {
			if isFirst || char != lastChar {
				normalized.WriteRune(char)
				lastChar = char
				isFirst = false
			}
		}

		return normalized.String()
	})
}

// RemoveAsterisks removes text surrounded by asterisks and fixes any resulting extra spaces.
func (tn *TextNormalizer) RemoveAsterisks(input string) string {
	if input == "" {
		return input
	}

	// Remove text between asterisks completely
	re := regexp.MustCompile(`\*[^*]*\*`)
	result := re.ReplaceAllString(input, "")

	// Fix any resulting double spaces or more
	spaceRe := regexp.MustCompile(`\s{2,}`)
	result = spaceRe.ReplaceAllString(result, " ")

	return result
}

// NormalizeCasing normalizes capitalization of words.
func (tn *TextNormalizer) NormalizeCasing(input string) string {
	if input == "" {
		return input
	}

	// Define word boundaries to identify words
	re := regexp.MustCompile(`\b\w+\b`)

	return re.ReplaceAllStringFunc(input, func(word string) string {
		// If the word has mixed capitalization
		if tn.hasMixedCasing(word) {
			// Convert everything to lowercase
			normalized := strings.ToLower(word)

			// First letter uppercase if the original was uppercase
			if len(normalized) > 0 && unicode.IsUpper([]rune(word)[0]) {
				runes := []rune(normalized)
				runes[0] = unicode.ToUpper(runes[0])
				normalized = string(runes)
			}

			return normalized
		}

		return word
	})
}

// hasMixedCasing checks if a word contains mixed capitalization.
func (tn *TextNormalizer) hasMixedCasing(word string) bool {
	if len(word) <= 1 {
		return false
	}

	hasUpper := false
	hasLower := false

	for _, char := range word {
		if unicode.IsUpper(char) {
			hasUpper = true
		} else if unicode.IsLower(char) {
			hasLower = true
		}

		if hasUpper && hasLower {
			return true
		}
	}

	return false
}

File: swift/TextNormalizer.swift
import Foundation

/// A class for normalizing text containing unusual writing styles,
/// such as repeated characters, text in asterisks, or
/// unusual capitalization.
public class TextNormalizer {
    
    /// Creates a new instance of TextNormalizer.
    public init() {}
    
    /// Normalizes text by applying multiple normalization rules.
    ///
    /// - Parameter input: The text to normalize.
    /// - Returns: The normalized text.
    public func normalizeText(_ input: String) -> String {
        if input.isEmpty {
            return input
        }
        
        var result = input
        
        // Normalize stretched words (e.g., "hellooo" -> "hello")
        result = normalizeRepeatedCharacters(result)
        
        // Remove text in asterisks (e.g., "*cough*")
        result = removeAsterisks(result)
        
        // Normalize capitalization (e.g., "HeLLo" -> "Hello")
        result = normalizeCasing(result)
        
        return result
    }
    
    /// Normalizes repeated characters in text.
    ///
    /// - Parameter input: The text to normalize.
    /// - Returns: The text with normalized repeated characters.
    public func normalizeRepeatedCharacters(_ input: String) -> String {
        if input.isEmpty {
            return input
        }
        
        do {
            // Define word boundaries to identify words
            let regex = try NSRegularExpression(pattern: "\\b\\w+\\b", options: [])
            let nsString = input as NSString
            let range = NSRange(location: 0, length: nsString.length)
            
            var lastEnd = 0
            var result = ""
            
            let matches = regex.matches(in: input, options: [], range: range)
            for match in matches {
                let wordRange = match.range
                let word = nsString.substring(with: wordRange)
                
                // Add text before the current word
                if wordRange.location > lastEnd {
                    result += nsString.substring(with: NSRange(location: lastEnd, length: wordRange.location - lastEnd))
                }
                
                // Normalize the word
                var normalized = ""
                var lastChar: Character?
                
                for char in word {
                    if lastChar != char {
                        normalized.append(char)
                        lastChar = char
                    }
                }
                
                result += normalized
                lastEnd = wordRange.location + wordRange.length
            }
            
            // Add remaining text
            if lastEnd < nsString.length {
                result += nsString.substring(with: NSRange(location: lastEnd, length: nsString.length - lastEnd))
            }
            
            return result
        } catch {
            // If regex fails, return the original input
            return input
        }
    }
    
    /// Removes text surrounded by asterisks and fixes any resulting extra spaces.
    ///
    /// - Parameter input: The text to normalize.
    /// - Returns: The text without asterisk content and no extra spaces.
    public func removeAsterisks(_ input: String) -> String {
        if input.isEmpty {
            return input
        }
        
        do {
            // Remove text between asterisks completely
            let regex = try NSRegularExpression(pattern: "\\*[^*]*\\*", options: [])
            let nsString = input as NSString
            let range = NSRange(location: 0, length: nsString.length)
            var result = regex.stringByReplacingMatches(in: input, options: [], range: range, withTemplate: "")
            
            // Fix any resulting double spaces or more
            let spaceRegex = try NSRegularExpression(pattern: "\\s{2,}", options: [])
            let resultNSString = result as NSString
            result = spaceRegex.stringByReplacingMatches(in: result, options: [], range: NSRange(location: 0, length: resultNSString.length), withTemplate: " ")
            
            return result
        } catch {
            // If regex fails, return the original input
            return input
        }
    }
    
    /// Normalizes capitalization of words.
    ///
    /// - Parameter input: The text to normalize.
    /// - Returns: The text with normalized capitalization.
    public func normalizeCasing(_ input: String) -> String {
        if input.isEmpty {
            return input
        }
        
        do {
            // Define word boundaries to identify words
            let regex = try NSRegularExpression(pattern: "\\b\\w+\\b", options: [])
            let nsString = input as NSString
            let range = NSRange(location: 0, length: nsString.length)
            
            var lastEnd = 0
            var result = ""
            
            let matches = regex.matches(in: input, options: [], range: range)
            for match in matches {
                let wordRange = match.range
                let word = nsString.substring(with: wordRange)
                
                // Add text before the current word
                if wordRange.location > lastEnd {
                    result += nsString.substring(with: NSRange(location: lastEnd, length: wordRange.location - lastEnd))
                }
                
                // If the word has mixed capitalization
                if hasMixedCasing(word) {
                    // Convert everything to lowercase
                    var normalized = word.lowercased()
                    
                    // First letter uppercase if the original was uppercase
                    if let firstChar = word.first, firstChar.isUppercase {
                        normalized = normalized.prefix(1).uppercased() + normalized.dropFirst()
                    }
                    
                    result += normalized
                } else {
                    result += word
                }
                
                lastEnd = wordRange.location + wordRange.length
            }
            
            // Add remaining text
            if lastEnd < nsString.length {
                result += nsString.substring(with: NSRange(location: lastEnd, length: nsString.length - lastEnd))
            }
            
            return result
        } catch {
            // If regex fails, return the original input
            return input
        }
    }
    
    /// Checks if a word contains mixed capitalization.
    ///
    /// - Parameter word: The word to check.
    /// - Returns: True if the word has mixed capitalization, otherwise False.
    private func hasMixedCasing(_ word: String) -> Bool {
        if word.count <= 1 {
            return false
        }
        
        var hasUpper = false
        var hasLower = false
        
        for char in word {
            if char.isUppercase {
                hasUpper = true
            } else if char.isLowercase {
                hasLower = true
            }
            
            if hasUpper && hasLower {
                return true
            }
        }
        
        return false
    }
}

// Example usage
/*
let normalizer = TextNormalizer()
let normalized = normalizer.normalizeText("Hellooo how areee you *cough* today?")
print(normalized)  // "Hello how are you today?"
*/